### DISCUSSION TITLE
Proxmox 7.3.4 LXC Container with docker

### DISCUSSION THREAD

**Discussion by netsrot303** [tteck/Proxmox#869] [Help Needed]
Since the last Proxmox update (7.3.4) I have problems with the LXC container with Docker. In the LXC container the created docker containers do not run anymore. Does anyone else have these problems? In the forum I read this: 
https://forum.proxmox.com/threads/todays-kernel-firmware-update-has-really-messed-up-my-boxes.119933/#post-520896

**Comment by tteck** [comment]
What file system are you running?

**Reply by netsrot303** [reply]
On Proxmox host runs zfs

**Reply by tteck** [reply]
I run ext4 so I can't troubleshoot. Did you try `fuse-overlayfs` like recommended in the post you referred to?

**Reply by netsrot303** [reply]
This fix worked for me only with a privileged container.
I'll test your new script in a moment.

**Comment by tteck** [comment]
If you'd like to test Docker with the option to use `fuse_overlayfs` (Advanced)

```
bash -c "$(wget -qLO - https://github.com/tteck/Proxmox/raw/main/ct/docker-v4.sh)"
```
@lordratner

**Reply by netsrot303** [reply]
The install scrippt fails.
++ curl -sSL https://get.docker.com
++ EXIT=100
++ LINE=116
++ error_exit
++ trap - ERR
++ local 'reason=Unknown failure occurred.'
++ local 'msg=Unknown failure occurred.'
++ local 'flag=\033[01;31m‼ ERROR \033[m100@116'
++ echo -e '\033[01;31m‼ ERROR \033[m100@116 Unknown failure occurred.'
‼ ERROR 100@116 Unknown failure occurred.
++ exit 100

**Reply by tteck** [reply]
OK, made some changes. If it works, check (`docker info | grep "Storage Driver"`) to make sure it's using `fuse-overlayfs` as the driver

**Reply by netsrot303** [reply]
The install script fails on installing docker. I tested privileged and unprivileged, both fails on installing docker

**Reply by tteck** [reply]
Sorry, I've been making changes every minute or so. I just made my last change for awhile, installs on my end now. Leave me your results and I'll get back on it (if needed) later. Thank you for helping me find a solution!

**Comment by netsrot303** [comment]
Now the install script finished and fuse-overlyfs is set as driver. portainer is running too. "docker pull jc21/nginx-proxy-manager" fails inside the lxc container with this error message "failed to register layer: ApplyLayer exit status 1 stdout:  stderr: unlinkat /var/cache/debconf: invalid argument". I tested this with an unprivileged Container. Later i will test this with a privileged LXC Container.

**Reply by tteck** [reply]
I guess we can call it _some_ progress 😜

**Comment by netsrot303** [comment]
So, I have now done the same with a privileged container and there I was able to pull the docker image from the NGINXproxymanager. The problem is only with an unprivileged container that you can't create some docker containers.

**Reply by tteck** [reply]
What kernel are you using?

**Comment by netsrot303** [comment]
Linux pve 5.15.83-1-pve #1 SMP PVE 5.15.83-1 (2022-12-15T00:00Z) x86_64 GNU/Linux

**Comment by tteck** [comment]
I've been using (`uname -r`) 6.1.0-1-pve I'm wondering if people running newer kernels are having this issue?

https://forum.proxmox.com/threads/opt-in-linux-6-1-kernel-for-proxmox-ve-7-x-available.119483/

I'm trying to narrow down what PVE 7.3-4 changed to cause this. kernel? zfs? ect... If we knew that, I'm guessing that a permanent fix would be much easier.

**Reply by tteck** [reply]
Seems some on newer kernels are having issues. So, it must be something with zfs 🤷🏻‍♂️

**Reply by netsrot303** [reply]
I will try it with the 6 kernel. Is it possible to downgrade again if I have problems?

**Reply by tteck** [reply]
yeah, but that may cause more issues with zfs
I'd just hold tight for now

**Reply by netsrot303** [reply]
Okay then it is probably better if I wait.
Thanks for help

**Reply by norbertjoni** [reply]
> Okay then it is probably better if I wait. Thanks for help

I had the same problem. I just reinstalled my containers with docker files and delete the old lxc.

**Reply by tteck** [reply]
What does that mean?

**Reply by norbertjoni** [reply]
I have to rebuild my dockers with a freshly installed LXC

**Reply by tteck** [reply]
Oh, I think @netsrot303 tried that. It only works with a privileged LXC

**Reply by rallegade** [reply]
Just to add to the conversation.

I was on 7.3-3 yesterday and tested the new addition with overlayer-fuse on ZFS and it worked perfectly fine.
Updated to 7.3-4 today to try it with the newer version and now I can only make new LXC's work with docker if they are privileged.
My old LXC running docker inside doesn't seem to be affected yet but they also have not been restarted... This is a bit of a shitshow and more users are reporting the same issue per this thread https://forum.proxmox.com/threads/docker-failed-to-register-layer-applylayer-exit-status-1-stdout-stderr-unlinkat-var-log-apt-invalid-argument.119954/

**Reply by rallegade** [reply]
Just to add to the confusion.

I have tested some minor things which seemed odd to me as any new LXC i made with your script is not working properly with fuse-overlayfs unless the LXC is privileged.
The old LXC i have which is still working is based on ubuntu and not debian like the ones from your script. Tested if this had anything to do with the fact that it still worked and the other ones do not and to my surprise any new LXC based on ubuntu running fuse-overlayfs is working perfectly fine for me without privileged mode enabled.

Could anyone else who have this issue please test it out as well?

**Reply by tteck** [reply]
What Ubuntu version are you using?

**Reply by tteck** [reply]
Docker on Ubuntu 22.04

```
bash -c "$(wget -qLO - https://github.com/tteck/Proxmox/raw/dev/ct/docker-v4.sh)"
```

**Reply by rallegade** [reply]
22.10

**Reply by rallegade** [reply]
Gonna test it straight away.

**Reply by rallegade** [reply]
did not work.

It might not be that it is using ubuntu as a base layer which is breaking it.
I tried editing  /etc/docker/daemon.json by removing the log journaled comment in there and adding

```
{
  "storage-driver": "fuse-overlayfs"
}
```

Then rebooting the LXC and trying docker compose up works.

This seems to be a problem with the fact that docker uses overlay2 as the default storage driver on Ubuntu and Debian as there is a folder in /var/lib/docker with this name present after the default installation with your script.

it should work per say, but looking at the section Supported backing filesystems 
in the documentation here https://docs.docker.com/storage/storagedriver/select-storage-driver/
only fuse-overlayfs supports any filesystem 


<p>With regard to Docker, the backing filesystem is the filesystem where
<code class="language-plaintext highlighter-rouge">/var/lib/docker/</code> is located. Some storage drivers only work with specific
backing filesystems.</p>



Storage driver | Supported backing filesystems
-- | --
overlay2, overlay | xfs with ftype=1, ext4
fuse-overlayfs | any filesystem
aufs | xfs, ext4
devicemapper | direct-lvm
btrfs | btrfs
zfs | zfs
vfs | any filesystem


EDIT: of course VFS does support any filesystem too but have shown problems in tandem with ZFS.

**Reply by tteck** [reply]
Without editing `/etc/docker/daemon.json` you can check which storage driver Docker is using `docker info | grep "Storage Driver"` the script doesn't use the `apt package` of `fuse-overlayfs`, it downloads from the source and puts it in `/usr/local/bin` where Docker should detect it during install. The updated ZFS is the issue.

**Reply by rallegade** [reply]
Well i can tell you that much that it absolutely does use overlay2 after a fresh install with the script, which for me at least is causing issues with docker.
As soon as i change the driver to `fuse-overlayfs` in  `/etc/docker/daemon.json` the problem is fixed.
Added some pics of it.

Output of `docker info | grep "Storage Driver"` straight after setup with the script
![overlay2](https://user-images.githubusercontent.com/200184/209829654-e282a81a-4e80-47c0-9259-69f98f3cbf2b.JPG)

Resulting in the following error
![error](https://user-images.githubusercontent.com/200184/209830553-3a1aa6d2-57fa-42e6-ac7a-25538fd8478b.JPG)

after changing to `fuse-overlayfs` in  `/etc/docker/daemon.json` 
![fuse-overlayfs](https://user-images.githubusercontent.com/200184/209830936-9323b79f-049f-4f87-b54d-08f41bf1fd2e.JPG)

so to me the easiest fix seems to be forcing docker to use `fuse-overlayfs`  by adding it to `/etc/docker/daemon.json`

**Reply by tteck** [reply]
Would you test the `dev` Docker LXC now?

**Reply by rallegade** [reply]
sure

**Reply by rallegade** [reply]
![error-setup](https://user-images.githubusercontent.com/200184/209838275-34ae515c-f615-4fc8-8a99-7f1cdcf1eaac.JPG)

**Reply by rallegade** [reply]
seems like you have a typo in the `/etc/docker/daemon.json` there is missing a `,`

it should be as the following

```
{
  "log-driver": "journald",
  "storage-driver": "fuse-overlayfs"
}
```

**Comment by netsrot303** [comment]
should i try the dev branch or the main branch?

**Reply by tteck** [reply]
either, dev is Ubuntu main is Debian
everything else is the same

**Reply by netsrot303** [reply]
ah okay, so for both only privileged containers works?

**Comment by lordratner** [comment]
OK, so I just spun up a V4 prod (debian) LXC. Unprivileged, fuse-overlayfs. It's all working great. I'm only running SWAG and Authelia on it for now, but no issues.

**Reply by netsrot303** [reply]
On host system you have zfs? can you pull the docker image "jc21/nginx-proxy-manager:latest" without errors?

**Reply by rallegade** [reply]
I was able to yes

**Reply by lordratner** [reply]
> On host system you have zfs? can you pull the docker image "jc21/nginx-proxy-manager:latest" without errors?

Yes, I'm using zfs. I pulled SWAG, which includes nginx. There were no issues
